File: platformio/servo/servo_py/servo_py/drive.py
import math

class Drive:
    def __init__(self, pwm_driver):
        pwm_driver.channels[0].duty_cycle = 0x0
        pwm_driver.channels[1].duty_cycle = 0x0
        self._pwm_driver = pwm_driver

    def set_throttle(self, throttle: float):
        if math.isfinite(throttle):
            duty_cycle = int(min(math.fabs(throttle), 1.0) * 0xffff)

            # First reset the PWM channel that is idle,
            # then set the PWM channel that drives in the desired direction.
            if throttle > 0.:
                self._pwm_driver.channels[1].duty_cycle = 0
                self._pwm_driver.channels[0].duty_cycle = duty_cycle
            else:
                self._pwm_driver.channels[0].duty_cycle = 0
                self._pwm_driver.channels[1].duty_cycle = duty_cycle

File: platformio/servo/servo/src/Drive.cpp
#include "Drive.hpp"

#include <utility>
#include <cmath>


namespace servo
{
    Drive::Drive(device_kit::pca9685::Pca9685&& pwm)
    :   pwm_ {std::move(pwm)}
    {
    }


    void Drive::throttle(float value)
    {
        if (std::isfinite(value))
        {
            std::uint16_t const duty_cycle = static_cast<std::uint16_t>(
                std::min(std::abs(value), 1.f) * pwm_.countsPerCycle());

            // First reset the PWM channel that is idle,
            // then set the PWM channel that drives in the desired direction.
            if (value > 0.f)
            {
                dutyCycle(0, 0);
                dutyCycle(1, duty_cycle);
            }
            else
            {
                dutyCycle(1, 0);
                dutyCycle(0, duty_cycle);
            }
        }
    }


    void Drive::dutyCycle(unsigned channel, std::uint16_t value)
    {
        pwm_.dutyCycle(channel, value > 0 ? 0 : pwm_.countsPerCycle(), value);
    }
}